redis/src/aio/connection.rs

```diff
@@ -256,108 +267,205 @@ where
 
     /// Subscribes to a new channel.
     pub async fn subscribe<T: ToRedisArgs>(&mut self, channel: T) -> RedisResult<()> {
-        cmd("SUBSCRIBE").arg(channel).query_async(&mut self.0).await
+        self.command(cmd("SUBSCRIBE").arg(channel)).await
     }
 
     /// Subscribes to a new channel with a pattern.
     pub async fn psubscribe<T: ToRedisArgs>(&mut self, pchannel: T) -> RedisResult<()> {
-        cmd("PSUBSCRIBE")
-            .arg(pchannel)
-            .query_async(&mut self.0)
-            .await
+        self.command(cmd("PSUBSCRIBE").arg(pchannel)).await
     }
 
     /// Unsubscribes from a channel.
     pub async fn unsubscribe<T: ToRedisArgs>(&mut self, channel: T) -> RedisResult<()> {
-        cmd("UNSUBSCRIBE")
-            .arg(channel)
-            .query_async(&mut self.0)
-            .await
+        self.command(cmd("UNSUBSCRIBE").arg(channel)).await
     }
 
     /// Unsubscribes from a channel with a pattern.
     pub async fn punsubscribe<T: ToRedisArgs>(&mut self, pchannel: T) -> RedisResult<()> {
-        cmd("PUNSUBSCRIBE")
-            .arg(pchannel)
-            .query_async(&mut self.0)
+        self.command(cmd("PUNSUBSCRIBE").arg(pchannel)).await
+    }
+}
+
+/// Represents a `PubSub` connection.
+pub struct SplitPubSubSink<C = Pin<Box<dyn AsyncStream + Send + Sync>>> {
+    sink: SplitSink<Framed<C, ValueCodec>, Vec<u8>>,
+    resp_recv: Receiver<Value>,
+}
+
+impl<C> SplitPubSubSink<C>
+where
+    C: Unpin + AsyncRead + AsyncWrite + Send,
+{
+    // Create a [`PubSubSink`] from a [`Connection`]
+    fn new(
+        framed: Framed<C, ValueCodec>,
+    ) -> (
+        SplitPubSubSink<C>,
+        impl Stream<Item = Result<Option<Msg>, RedisError>>,
+    ) {
+        let (sink, stream) = framed.split();
+        let (resp_send, resp_recv) = channel::<Value>(1);
+
+        (
+            Self { sink, resp_recv },
+            unfold((stream, resp_send), |(mut stream, mut resp_send)| {
+                Box::pin(async move {
+                    let item = stream.next().await;
+
+                    match item {
+                        Some(Ok(Ok(value))) => Some(Ok(match Msg::from_value(&value) {
+                            Some(msg) => Some(msg),
+                            _ => {
+                                let _ = resp_send.send(value).await;
+
+                                None
+                            }
+                        })),
+                        _ => None,
+                    }
+                    .map(|res| (res, (stream, resp_send)))
+                })
+            }),
+        )
+    }
+
+    // Deliver the PUBSUB command to this PUBSUB connection
+    async fn command<'a>(&'a mut self, cmd: &'a Cmd) -> RedisResult<()> {
+        let mut out = Vec::new();
+
+        cmd.write_packed_command(&mut out);
+        self.sink.send(out).await?;
+
+        self.resp_recv
+            .next()
             .await
+            .map(|_| ())
+            .ok_or(RedisError::from((
```

Comment:
you're ignoring here errors that might be relevant - for example, SSUBSCRIBE (which isn't implemented here, but this might be the basis for a cluster pubsub) can return MOVED.